security/pkg/server/ca/server_test.go

```diff
@@ -70,6 +76,130 @@ func (authn *mockAuthenticator) Authenticate(ctx context.Context) (*authenticate
 	}, nil
 }
 
+type mockAuthInfo struct {
+	authType string
+}
+
+func (ai mockAuthInfo) AuthType() string {
+	return ai.authType
+}
+
+func TestCreateCertificateE2EWithCertificates(t *testing.T) {
+	callerID := "test.identity"
+	ids := []util.Identity{
+		{Type: util.TypeURI, Value: []byte(callerID)},
+	}
+	sanExt, err := util.BuildSANExtension(ids)
+	if err != nil {
+		t.Error(err)
+	}
+	auth := &authenticate.ClientCertAuthenticator{}
+
+	server := &Server{
+		ca: &mockca.FakeCA{
+			SignedCert: []byte("cert"),
+			KeyCertBundle: &mockutil.FakeKeyCertBundle{
+				CertChainBytes: []byte("cert_chain"),
+				RootCertBytes:  []byte("root_cert"),
+			},
+		},
+		hostnames:      []string{"hostname"},
+		port:           8080,
+		Authenticators: []authenticate.Authenticator{auth},
+		monitoring:     newMonitoringMetrics(),
+	}
+	mockCertChain := []string{"cert", "cert_chain", "root_cert"}
+	testCerts := map[string]struct {
+		certChain          [][]*x509.Certificate
+		caller             *authenticate.Caller
+		authenticateErrMsg string
```

Comment:
Please verify authenticateErrMsg in the for loop.

Reply:
Yes I try to add this . however from the code(https://github.com/williamaronli/istio/blob/30bc6aec6b333d6d51c4279e2629337cace11b05/security/pkg/server/ca/server.go#L93-L97) it will only return a "request authenticate failure" error
```
caller := s.authenticate(ctx)
	if caller == nil {
		s.monitoring.AuthnError.Increment()
		return nil, status.Error(codes.Unauthenticated, "request authenticate failure")
	}
```
the authenticateErrMsg error message defined in the test case will be show in the warning:
https://github.com/williamaronli/istio/blob/30bc6aec6b333d6d51c4279e2629337cace11b05/security/pkg/server/ca/server.go#L297
like below(it is from my glinux):

2020-06-27T05:36:48.286144Z	warn	serverca	Authentication failed for 192.168.1.1: Authenticator ClientCertAuthenticator at index 0 got error: unsupported auth type: "not-tls". 
2020-06-27T05:36:48.286212Z	warn	serverca	Authentication failed for 192.168.1.1: Authenticator ClientCertAuthenticator at index 0 got error: no verified chain is found. 
2020-06-27T05:36:48.286219Z	warn	serverca	Authentication failed for 192.168.1.1: Authenticator ClientCertAuthenticator at index 0 got error: the SAN extension does not exist. 
2020-06-27T05:36:48.286236Z	warn	serverca	Authentication failed for unknown: Authenticator ClientCertAuthenticator at index 0 got error: no client certificate is presented

And in the cert_authentication_test.go(https://github.com/williamaronli/istio/blob/caf1798d024ec94369359fe924dc8d81da18f7b7/security/pkg/server/ca/authenticate/cert_authenticator_test.go#L40)
it already check the content of the error message. So I think no need to do it once again here

Reply:
In that case, could you just replace the authenticateErrMsg with a comment for each case describing the error?